oasislmf/utils/data.py

```diff
@@ -1042,3 +1048,41 @@ def set_dataframe_column_dtypes(df, dtypes):
     df = df.astype(_dtypes)
 
     return df
+
+
+def fill_na_with_categoricals(df, fill_value):
+    """
+    Fill NA values in a Pandas DataFrame, with handling for Categorical dtype columns.
+
+    The input dataframe is modified inplace.
+
+    :param df: The dataframe to process
+    :type df: pd.DataFrame
+
+    :param fill_value: A single value to use in all columns, or a dict of column names and
+                       corresponding values to fill.
+    :type fill_value: int, float, str, dict
+
+    """
+    if not isinstance(fill_value, dict):
+        fill_value = {col_name: fill_value for col_name in df.columns}
+
+    for col_name, value in fill_value.items():
+        if col_name not in df:
+            continue
+
+        col = df[col_name]
+        if pd.api.types.is_categorical_dtype(col):
+            # Force to be a string - using categorical for string columns
+            value = str(value)
+            fill_value[col_name] = value
+
+            if value not in col.cat.categories:
+                fill_value[col_name] = value
+                col.cat.add_categories([value], inplace=True)
+
+    # Note bug in Pandas 1.1.0/1.1.1 for fillna(inplace=True) when categorical dtypes are involved
+    # It doesn't hit here, but leads to very confusing errors internal to Pandas later on
+    # https://github.com/pandas-dev/pandas/issues/35731
+    df.replace('nan', np.nan, inplace=True)
```

Comment:
Sorry, not sure to understand this specific case. 'nan' should be evaluated as NaN=np.nan during the loading and the issue mentioned shouldn't be present as you exclude the bugged version.

Reply:
This one was brought up by the Oasis tests. When using 'categorical' as the data type, it seems that Pandas _doesn't_ automatically do that mapping. The test was therefore inputting 'nan' as a string, expecting NaN back, but was actually getting 'nan' again.

The comment about the bug is separate to the actual code (but boy, did it make debugging the problem fun) - I'll make it all a little clearer.

Reply:
(As with my other comment, getting Pandas not to map 'nan' again seems to have been an awkward edge case, as a simple CSV test actually shows it does convert as expected. I'll work out the exact cause and note it down!)

Reply:
As with the other section, I can't reproduce the original issue I was trying to fix - it may have been part of the confusion caused by that Pandas bug.